coqPackages: etc

CohenCyril opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Many backward compatible changes
- additional format `owner/repo:branch` to fetch sources in meta-fetcher
- new argument `origin` to `coq` and `mkCoqDerivation` in order to
  replace the current input `version` attribute, so that ultimately
  the input and output `version` coincide. To preserve backward compatibility
  we still allow to use `version` instead of `origin` but we send a warning
  if the input and output versions differ. (This might be turned into an
  error in the future).
  + Refactored `coqPackages` to pass it to every derivation,
- `mkCoqDerivation` can now take a function instead of an attribute set,
  the argument is a previous iteration of the arguments that `mkCoqDerivation`
  pass to `mkDerivation`, mostly to get the computed version.
  + simplified derivation of `mathcomp-analysis` thanks to this.

All changes are duely documented


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

CohenCyril commented:
@GrahamcOfBorg build coqPackages_8_11 coqPackages_8_12 coqPackages_8_13 coqPackages_8_10

CohenCyril commented:
I do not understand this error
```
/nix/store/99x7iwwqs1c29qcw9hm489pc70fferror: access to URI 'https://gitlab.mpi-sws.org/iris/iris/-/archive/3.3.0/iris-3.3.0.tar.gz' is forbidden in restricted mode
```
@vbgl @siraben do you understand this?

Zimmi48 commented:
Some comments:
- `origin` is not a very informative name. `version` sounded better, but I do understand the problem. Can we think of a better name?
- You also changed the version of non-release origins to `999` (instead of `dev`). Did you actually hit a problem because of the `dev` version?
- `owner@repo:branch` is not very standard syntax. Is there an issue with the much more natural `owner/repo:branch`?

Zimmi48 commented on pkgs/applications/science/logic/coq/default.nix:
```diff
@@ -48,27 +48,28 @@ let
+      location = { owner = "coq"; repo = "coq";}; }
+    { inherit origin version; pname = "coq"; };
+  coq-version = args.coq-version or (versions.majorMinor fetched.version);
+  coqAtLeast = lib.versionAtLeast coq-version;
```
`coqAtLeast` should probably get documented as well.

CohenCyril replied:
actually it is local to coq/default.nix so I'm not sure it should...

CohenCyril commented:
> * `origin` is not a very informative name. `version` sounded better, but I do understand the problem. Can we think of a better name?

That's the best name I could come with: `src` or any already used attribute name is excluded, and variations that are too close (like `source` for example) would be clumsy IMO. I'm open to suggestions though.

> * You also changed the version of non-release origins to `999` (instead of `dev`). Did you actually hit a problem because of the `dev` version?

Yes: it would be convenient that version comparison operations do satisfy `x.y.z < dev` (where `x`, `y` and `z` are numbers`. So I thought `999` both satisfied this condition and somehow intuitively mean "dev"...

> * `owner@repo:branch` is not very standard syntax. Is there an issue with the much more natural `owner/repo:branch`?

I cannot remember, I thought it could be ambiguous but I cannot remember the source of ambiguity...

CohenCyril commented on pkgs/applications/science/logic/coq/default.nix:
```diff
@@ -48,27 +48,28 @@ let
-    args.version;
-  version = fetched.version;
-  coq-version = args.coq-version or (if version != "dev" then versions.majorMinor version else "dev");
-  versionAtLeast = v: (coq-version == "dev") || (lib.versionAtLeast coq-version v);
```
@Zimmi48 this is one of the irregularities of using `dev` as the `version`

CohenCyril committed:
coqPackages: etc

Many backward compatible changes
- additional format "owner@repo:branch" to fetch sources in meta-fetcher
- new argument `origin` to `coq` and `mkCoqDerivation` in order to
  replace the current input `version attribute, so that ultimately
  the input and output `version` coincide. To preserve backward compatibility
  we still allow to use `version` instead of `origin` but we send a warning
  if the input and output versions differ. (This might be turned into an
  error in the future).
  + Refactored `coqPackages` to pass it to every derivation,
- `mkCoqDerivation` can now take a function instead of an attribute set,
  the argument is a previous iteration of the arguments that `mkCoqDerivation`
  pass to `mkDerivation`, mostly to get the computed version.
  + simplified derivation of `mathcomp-analysis` thanks to this.

All changes are duely documented

Zimmi48 commented:
> > * `origin` is not a very informative name. `version` sounded better, but I do understand the problem. Can we think of a better name?
> 
> That's the best name I could come with: `src` or any already used attribute name is excluded, and variations that are too close (like `source` for example) would be clumsy IMO. I'm open to suggestions though.

What about `sourceOverride`? Not very simple but at least the purpose is clear.

CohenCyril commented:
> > > * `origin` is not a very informative name. `version` sounded better, but I do understand the problem. Can we think of a better name?
> > 
> > 
> > That's the best name I could come with: `src` or any already used attribute name is excluded, and variations that are too close (like `source` for example) would be clumsy IMO. I'm open to suggestions though.
> 
> What about `sourceOverride`? Not very simple but at least the purpose is clear.

I don't like it... (too verbose, and not clear to me).
I'm thinking about `shortSrc` or `fixSrc`...

Zimmi48 commented:
OK, well, let's keep `origin` if we cannot find anything better.

Zimmi48 commented on pkgs/build-support/coq/default.nix:
```diff
@@ -4,106 +4,112 @@ with builtins; with lib;
+  // (optionalAttrs useDune2 {
+    installPhase = ''
+      runHook preInstall
+      dune install --prefix=$out
```
This is unrelated to this PR, but this is actually incorrect in case of a monorepo. Anyway, `useDune2` is also broken for other reasons (it doesn't override the configure and build phases).

Zimmi48 replied:
This was fixed in another PR.

Zimmi48 commented on pkgs/development/coq-modules/autosubst/default.nix:
```diff
@@ -1,15 +1,18 @@
 mkCoqDerivation {
-  pname = "autosubst";
+  pname   = "autosubst";
+  owner   = "uds-psl";
```
This repo redirects to the coq-community one, so this addition is not useful.

CohenCyril replied:
```suggestion
```

Zimmi48 commented on pkgs/development/coq-modules/autosubst/default.nix:
```diff
@@ -1,15 +1,18 @@
   defaultVersion = with versions; switch coq.coq-version [
-    { case = isGe "8.10"; out = "1.7"; }
+    { case = isGe "8.10";       out = "1.7"; }
+    { case = range "8.5" "8.7"; out = "5b40a32e"; }
```
This is a commit outside the main repository. Are you sure about this?

CohenCyril replied:
no

Zimmi48 commented on pkgs/top-level/coq-packages.nix:
```diff
@@ -97,8 +97,8 @@ let
-  mkCoq = version: callPackage ../applications/science/logic/coq {
-    inherit version ocamlPackages_4_05 ocamlPackages_4_09 ocamlPackages_4_10;
+  mkCoq = origin: callPackage ../applications/science/logic/coq {
+    inherit origin ocamlPackages_4_05 ocamlPackages_4_09 ocamlPackages_4_10;
```
This is a case where `version` would be just as fine, wouldn't it?

Zimmi48 commented on doc/languages-frameworks/coq.section.md:
````diff
@@ -41,10 +45,18 @@ The recommended way of defining a derivation for a Coq library, is to use the `c
 
+The builder `mkCoqDerivation` can also by applied to a function taking an attribute set and returning one. The main reason for using this feature is to get the version number as computed by `mkCoqDerivation`, as in:
+```
+cat $(nix-build -E 'with import ./. {}; coqPackages.mkCoqDerivation (computed: {
+  pname = "test"; src = mktemp; defaultVersion = "1.0";
+  VERSION = computed.version; installPhase = "echo $VERSION > $out";})')
+```
+will return `999`. This does a fixpoint computation, so do not use an attribute of `computed` in order to alter the outcome of the same attribute or you will get an infinite recursion.  For a concrete example see `coqPackage.mathcomp-analysis`.
````
The `mathcomp-analysis` example is way more interesting *and* understandable that this one. So I would suggest quoting it directly in the doc here (but skipping the parts that are not relevant to illustrate the behavior).

siraben commented:
Result of `nixpkgs-review pr 116690` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

Zimmi48 commented:
IMHO it's too bad that this PR has waited for so long. I think only the version number change `dev` -> `999` needs more thinking so should be removed from this PR and discussed separately. The rest is almost ready, I've already reviewed it (see comments above) and only minor changes (mostly in the doc) are needed.

Zimmi48 commented:
@CohenCyril As I said above, I think we should defer the version number change to a later PR and merge the rest. With respect to the version number, we should probably use what comes out of the discussion at https://github.com/NixOS/rfcs/pull/107. I like the idea of starting from an existing version number and adding a suffix (`+unstable`...) as it maps well with the possibility of taking a dev version for another branch. We could even chime in the RFC discussion if that's useful.

CohenCyril commented:
> @CohenCyril As I said above, I think we should defer the version number change to a later PR and merge the rest. With respect to the version number, we should probably use what comes out of the discussion at [NixOS/rfcs#107](https://github.com/NixOS/rfcs/pull/107). I like the idea of starting from an existing version number and adding a suffix (`+unstable`...) as it maps well with the possibility of taking a dev version for another branch. We could even chime in the RFC discussion if that's useful.

I agree, I just did not have the time to deal with this yet.

CohenCyril committed:
Apply suggestions from code review

Co-authored-by: Théo Zimmermann <theo.zimmi@gmail.com>
